lib/ldp.js

```diff
@@ -19,6 +19,160 @@ var doWhilst = require('async').doWhilst
+      (S(file).endsWith(turtleExtension) ? '' : ldp.suffixMeta)
+
+    getMetadataGraph(ldp, metaFile, baseUri, function (err, metadataGraph) {
+      if (err) {
+        metadataGraph = $rdf.graph()
+      }
+
+      // Add Container or BasicContainer types
+      if (stats.isDirectory()) {
+        resourceGraph.add(
+          metadataGraph.sym(fileSubject),
+          ns.rdf('type'),
+          ns.ldp('BasicContainer'))
+
+        resourceGraph.add(
+          metadataGraph.sym(fileSubject),
+          ns.rdf('type'),
+          ns.ldp('Container'))
+      }
+      // Add generic LDP type
+      resourceGraph.add(
+        metadataGraph.sym(fileSubject),
+        ns.rdf('type'),
+        ns.ldp('Resource'))
+
+      // Add type from metadataGraph
+      metadataGraph
+        .statementsMatching(
+          metadataGraph.sym(baseUri),
+          ns.rdf('type'),
+          undefined)
+        .forEach(function (typeStatement) {
+          // If the current is a file and its type is BasicContainer,
+          // This is not possible, so do not infer its type!
+          if (
+            (
+              typeStatement.object.uri !== ns.ldp('BasicContainer').uri &&
+              typeStatement.object.uri !== ns.ldp('Container').uri
+            ) ||
+            !stats.isFile()
+          ) {
+            resourceGraph.add(
+              resourceGraph.sym(fileSubject),
+              typeStatement.predicate,
+              typeStatement.object)
+          }
+        })
+
+      return callback(null)
+    })
+  })
+}
+
+function addStats (resourceGraph, baseUri, stats) {
+  resourceGraph.add(
+    resourceGraph.sym(baseUri),
+    ns.stat('mtime'),
+    stats.mtime.getTime() / 1000)
+
+  resourceGraph.add(
+    resourceGraph.sym(baseUri),
+    ns.stat('size'),
+    stats.size)
+}
+
+function readdir (filename, callback) {
+  debug.handlers('GET -- Reading directory')
+  fs.readdir(filename, function (err, files) {
+    if (err) {
+      debug.handlers('GET -- Error reading files: ' + err)
+      return callback(error(err, 'Can\'t read container'))
+    }
+
+    debug.handlers('Files in directory: ' + files)
+    return callback(null, files)
+  })
+}
+
+function getMetadataGraph (ldp, metaFile, fileBaseUri, callback) {
+  ldp.stat(metaFile, function (err, metaStats) {
+    if (err) {
+      return callback(err)
+    }
+
+    if (metaStats && metaStats.isFile()) {
+      ldp.readFile(metaFile, function (err, rawMetadata) {
+        if (err) {
+          return callback(err)
+        }
+
+        var metadataGraph = $rdf.graph()
+        try {
+          $rdf.parse(
+            rawMetadata,
+            metadataGraph,
+            fileBaseUri,
+            'text/turtle')
+        } catch (dirErr) {
+          return callback(error(err, 'Can\'t parse container metadata'))
+        }
+        return callback(null, metadataGraph)
+      })
+    } else {
+      return callback(null, $rdf.graph())
+    }
+  })
+}
+
```

Comment:
Defining multiple nested functions inside a function, and then calling them, is a pretty serious anti-pattern tho. 
Feel free to move em to a separate file or module tho.